### ISSUE TITLE
使用oc-softmax之后得分问题

### ISSUE THREAD

**Issue by beijita-yegucheng** [yzyouzhang/AIR-ASVspoof#19]
为什么使用oc-softmax之后得分，正样本的得分是再-1左右，而负样本是在1左右。但是使用softmax，得分确是正样本在1左右，负样本在0左右？

**Comment by yzyouzhang** [comment]
感谢您的提问！我们使用oc-softmax打分时候是用的cosine similarity，所以得分是在[-1,1]，我们的符合可能写反了，我会检查一下。softmax因为在二分类的时候就等同sigmoid，所以是[0,1]，是这样的

**Comment by yzyouzhang** [comment]
oc-softmax之后的得分不会放入softmax，请问放入softmax之前是怎样的输出？

**Comment by JJ-Guo1996** [comment]
输入softmax的数据不是ocsoftmax输出的分数，而是lfcc_outputs，如下所示。
feats, lfcc_outputs = self.model(lfcc)
out_score = F.softmax(lfcc_outputs)


Junjun Guo
***@***.***
发自 网易邮箱大师



---- 回复的原邮件 ----
发件人 You ***@***.***>日期 2022年05月18日 11:56收件人 ***@***.***>抄送至 ***@***.******@***.***>主题 Re: [yzyouzhang/AIR-ASVspoof] 使用oc-softmax之后得分问题 (#19)

oc-softmax之后的得分不会放入softmax，请问放入softmax之前是怎样的输出？

**Comment by yzyouzhang** [comment]
关于正确的给样本打分的方式请看[test.py](https://github.com/yzyouzhang/AIR-ASVspoof/blob/master/test.py)。

**Comment by JJ-Guo1996** [comment]
![1](https://user-images.githubusercontent.com/76087490/168962088-ef668b0c-5852-4b4e-b7b3-b193afceff3b.PNG)
标红的部分是用softmax打分吧！？但是结果就像我之前提到的。

**Comment by yzyouzhang** [comment]
在这个的两行之后这个score就被覆盖了

**Comment by JJ-Guo1996** [comment]
即使重新定义score为out_score 也是同样的效果呢。现在的问题就是在测试的时候，我想用oc-softmax训练好的模型通过softmax计算样本的分数，但是预测结果比较差！
郭军军
***@***.***


签名由 网易邮箱大师 定制


在2022年5月18日 13:21，You ***@***.***> 写道：


在这个的两行之后这个score就被覆盖了

**Comment by yzyouzhang** [comment]
用oc-softmax训练好的模型就应该通过ocsoftmax计算样本的分数 而不是softmax。
ocsoftmax里面的weight vector也是训练出来的，这些都在models文件夹里有保存

**Comment by yzyouzhang** [comment]
```
if add_loss == "ocsoftmax":
     ang_isoloss, score = loss_model(feats, labels)
```
应该是用这里的score

**Comment by JJ-Guo1996** [comment]
好的 明白了多谢！！！  另外有没有考虑对算法优化呢？比如数据增强或者模型融合？


郭军军
***@***.***


签名由 网易邮箱大师 定制


在2022年5月18日 14:02，You ***@***.***> 写道：


用oc-softmax训练好的模型就应该通过ocsoftmax计算样本的分数 而不是softmax。
 ocsoftmax里面的weight vector也是训练出来的，这些都在models文件夹里有保存

**Comment by yzyouzhang** [comment]
请参考我们的一些follow-up works:
[1] You Zhang, Ge Zhu, Fei Jiang, and Zhiyao Duan, ``An Empirical Study on Channel Effects for Synthetic Voice Spoofing Countermeasure Systems'', in Proc. Interspeech, pp. 4309-4313, 2021. [[link](https://www.isca-speech.org/archive/interspeech_2021/zhang21ea_interspeech.html)][[code](https://github.com/yzyouzhang/Empirical-Channel-CM)][[video](https://www.youtube.com/watch?v=vLijNUJklo0)]
[2] Xinhui Chen*, You Zhang*, Ge Zhu*, and Zhiyao Duan, ``UR Channel-Robust Synthetic Speech Detection System for ASVspoof 2021'', in Proc. ASVspoof 2021 Workshop, pp. 75-82, 2021. (* equal contribution) [[link](https://www.isca-speech.org/archive/asvspoof_2021/chen21_asvspoof.html)][[code](https://github.com/yzyouzhang/ASVspoof2021_AIR)][[video](https://www.youtube.com/watch?v=-wKMOTp8Tt0)]
[3] You Zhang, Fei Jiang, Ge Zhu, Xinhui Chen, and Zhiyao Duan, ``Generalizing Voice Presentation Attack Detection to Unseen Synthetic Attacks and Channel Variation'', Handbook of Biometric Anti-spoofing, Springer, 2022. (to be published) [[code](https://github.com/yzyouzhang/HBAS_chapter_voice3)]
[4] You Zhang, Ge Zhu, and Zhiyao Duan, ``A Probabilistic Fusion Framework for Spoofing Aware Speaker Verification'', in Proc. Odyssey, 2022. (accepted) [[link](https://arxiv.org/abs/2202.05253)][[code](https://github.com/yzyouzhang/SASV_PR)]
